x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/SearchHitCursor.java

```diff
@@ -0,0 +1,224 @@
+                ),
+                listener::onFailure
+            )
+        );
+    }
+
+    protected Supplier<SearchHitRowSet> makeRowSet(int sizeRequested, SearchResponse response) {
+        return () -> new SearchHitRowSet(extractors, mask, sizeRequested, limit, response);
+    }
+
+    static void handle(
+        Client client,
+        SearchResponse response,
+        SearchSourceBuilder source,
+        Supplier<SearchHitRowSet> makeRowSet,
+        ActionListener<Page> listener,
+        boolean includeFrozen
+    ) {
+
+        if (log.isTraceEnabled()) {
+            Querier.logSearchResponse(response, log);
+        }
+
+        SearchHit[] hits = response.getHits().getHits();
+
+        SearchHitRowSet rowSet = makeRowSet.get();
+
+        if (rowSet.hasRemaining() == false) {
+            Querier.closePointInTime(
+                client,
+                response.pointInTimeId(),
+                ActionListener.wrap(r -> listener.onResponse(Page.last(rowSet)), listener::onFailure)
+            );
+        } else {
+            source.pointInTimeBuilder(new PointInTimeBuilder(response.pointInTimeId()));
+            updateSearchAfter(hits, source);
+
+            byte[] nextQuery;
+            try {
+                nextQuery = Querier.serializeQuery(source);
+            } catch (IOException e) {
+                listener.onFailure(e);
+                return;
+            }
+
+            SearchHitCursor nextCursor = new SearchHitCursor(
+                nextQuery,
+                rowSet.extractors(),
+                rowSet.mask(),
+                rowSet.getRemainingLimit(),
+                includeFrozen
+            );
+            listener.onResponse(new Page(rowSet, nextCursor));
+        }
+    }
+
+    private static void updateSearchAfter(SearchHit[] hits, SearchSourceBuilder source) {
+        assert hits.length > 0;
+        SearchHit lastHit = hits[hits.length - 1];
+        source.searchAfter(lastHit.getSortValues());
+    }
+
+    @Override
+    public void clear(Client client, NamedWriteableRegistry registry, ActionListener<Boolean> listener) {
+        SearchSourceBuilder query;
+        try {
+            query = Querier.deserializeQuery(registry, nextQuery);
+        } catch (IOException e) {
+            listener.onFailure(e);
+            return;
+        }
+        Check.isTrue(query.pointInTimeBuilder() != null, "Expected cursor with point-in-time id but got null");
+        Querier.closePointInTime(client, query.pointInTimeBuilder().getEncodedId(), listener);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(Arrays.hashCode(nextQuery), extractors, limit, mask, includeFrozen);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null || obj.getClass() != getClass()) {
+            return false;
+        }
+        SearchHitCursor other = (SearchHitCursor) obj;
+        return Arrays.equals(nextQuery, other.nextQuery)
```

Comment:
I would do the "costly" comparisons last and have `includeFrozen` and `limit` as first ones.